### PR TITLE
Speed up checked kernels for non-null data (~1.4-5x faster)

### LINKED ISSUE #2747: Improve performance of checked kernels on non-null data
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
There is some low hanging fruit to improve the checked arithmic kernels.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Dandandan**
# Which issue does this PR close?

Closes #2747

# Rationale for this change
We can use some more optimized ways to generate new arrays from fallible iterators:
(There is an overflow in the benchmark).
 
```
add_checked(0)          time:   [14.555 µs 14.574 µs 14.596 µs]
                        change: [-79.379% -79.172% -78.968%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe

Benchmarking subtract_checked(0): Collecting 100 samples in estimated 5.0209 s (subtract_checked(0)     time:   [15.246 µs 15.265 µs 15.285 µs]
                        change: [-82.424% -81.720% -81.002%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  2 (2.00%) high severe

Benchmarking multiply_checked(0): Collecting 100 samples in estimated 5.0334 s (multiply_checked(0)     time:   [14.364 µs 14.381 µs 14.399 µs]
                        change: [-80.759% -80.340% -79.944%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) low mild
  4 (4.00%) high mild
  4 (4.00%) high severe

Benchmarking divide_checked(0): Collecting 100 samples in estimated 5.2102 s (96divide_checked(0)       time:   [54.127 µs 54.173 µs 54.223 µs]
                        change: [-29.294% -29.093% -28.909%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  5 (5.00%) high mild
  4 (4.00%) high severe
```

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/compute/kernels/arity.rs [hunk 1]
```diff
@@ -292,19 +292,16 @@ where
         .map(|x| len - x.count_set_bits())
         .unwrap_or_default();
 
-    let mut buffer = BufferBuilder::<O::Native>::new(len);
-    buffer.append_n_zeroed(len);
-    let slice = buffer.as_slice_mut();
+    let values = a.values().iter().zip(b.values()).map(|(l, r)| op(*l, *r));
 
-    try_for_each_valid_idx(len, 0, null_count, null_buffer.as_deref(), |idx| {
```
Comment (viirya) [inline arrow/src/compute/kernels/arity.rs:299]: `try_unary` also uses `try_for_each_valid_idx`, same trick also applies there?

**Comment by Dandandan** [comment]
Ah wait - this changes the semantics slightly as the overflow check probably isn't meant to work on null-values...
Requires some more thought.

File: arrow/src/compute/kernels/arity.rs (hunk 1 above)
Comment (Dandandan) [reply arrow/src/compute/kernels/arity.rs]: Yes - applied it there too, thanks.

**Review by tustvold — APPROVED** [review]
Makes sense, no point zero allocating if you don't need to 👍

**Review by viirya — APPROVED** [review]

**Pull request merged**